File: src/main/java/pl/mjedynak/idea/plugins/pit/actions/RunSomeTestsPitAction.java
package pl.mjedynak.idea.plugins.pit.actions;

import com.intellij.openapi.fileTypes.FileType;
import com.intellij.openapi.fileTypes.StdFileTypes;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiManager;
import org.jetbrains.annotations.NotNull;
import pl.mjedynak.idea.plugins.pit.cli.PitCommandLineArgumentsContainer;
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerFactory;
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerPopulator;
import pl.mjedynak.idea.plugins.pit.configuration.PitConfigurationType;
import pl.mjedynak.idea.plugins.pit.configuration.PitRunConfiguration;

import static pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_TESTS;

/**
 * Runs pitest filtered to a directory of test files
 */
public class RunSomeTestsPitAction extends DirectoryOrFilePitAction {

    boolean isEnabled(@NotNull final Project project, @NotNull final Module module,
                               @NotNull final VirtualFile vfile) {
        return (vfile.isDirectory() || (vfile.getFileType() == StdFileTypes.JAVA)) &&
                module.getModuleTestsWithDependentsScope().contains(vfile);
    }

    @Override
    String getTitleForItem(final String item) {
        return "Pitest using tests in '" + item + "'";
    }

    @Override
    PitRunConfiguration makeConfigurationForClassList(@NotNull final String classList, @NotNull final Project project, @NotNull final String title) {
        final DefaultArgumentsContainerPopulator defaultArgumentsContainerPopulator = new DefaultArgumentsContainerPopulator(
                ProjectRootManager.getInstance(project), PsiManager.getInstance(project));

        final DefaultArgumentsContainerFactory defaultArgumentsContainerFactory
                = new DefaultArgumentsContainerFactory(defaultArgumentsContainerPopulator) {
            @Override
            public PitCommandLineArgumentsContainer createDefaultPitCommandLineArgumentsContainer(final Project proj) {
                final PitCommandLineArgumentsContainer container = super.createDefaultPitCommandLineArgumentsContainer(proj);
                container.put(TARGET_TESTS, classList);
                return container;
            }
        };

        return new PitRunConfiguration("PIT using tests in " + title, project,
                PitConfigurationType.getInstance().getConfigurationFactories()[0],
                defaultArgumentsContainerFactory);

    }

}

File: src/main/kotlin/pl/mjedynak/idea/plugins/pit/actions/RunSomeTestsPitAction.kt
package pl.mjedynak.idea.plugins.pit.actions

import com.intellij.ide.highlighter.JavaFileType
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ProjectRootManager
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiManager
import pl.mjedynak.idea.plugins.pit.cli.PitCommandLineArgumentsContainer
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerFactory
import pl.mjedynak.idea.plugins.pit.cli.factory.DefaultArgumentsContainerPopulator
import pl.mjedynak.idea.plugins.pit.cli.model.PitCommandLineArgument.TARGET_TESTS
import pl.mjedynak.idea.plugins.pit.configuration.PitConfigurationType
import pl.mjedynak.idea.plugins.pit.configuration.PitRunConfiguration

class RunSomeTestsPitAction : DirectoryOrFilePitAction() {
    override fun isEnabled(
        project: Project,
        module: Module,
        vfile: VirtualFile,
    ): Boolean =
        (vfile.isDirectory || vfile.fileType == JavaFileType.INSTANCE) &&
            module.moduleTestsWithDependentsScope.contains(vfile)

    override fun getTitleForItem(item: String): String = "Pitest using tests in '$item'"

    override fun makeConfigurationForClassList(
        classList: String,
        project: Project,
        title: String,
    ): PitRunConfiguration {
        val defaultArgumentsContainerPopulator =
            DefaultArgumentsContainerPopulator(
                ProjectRootManager.getInstance(project),
                PsiManager.getInstance(project),
            )
        val defaultArgumentsContainerFactory =
            object : DefaultArgumentsContainerFactory(defaultArgumentsContainerPopulator) {
                override fun createDefaultPitCommandLineArgumentsContainer(`project`: Project): PitCommandLineArgumentsContainer {
                    val container = super.createDefaultPitCommandLineArgumentsContainer(project)
                    container.put(TARGET_TESTS, classList)
                    return container
                }
            }
        return PitRunConfiguration(
            "PIT using tests in $title",
            project,
            PitConfigurationType.getInstance()!!.configurationFactories[0],
            defaultArgumentsContainerFactory,
        )
    }
}
